src/bodygraph.cpp

```diff
@@ -0,0 +1,577 @@
+#include "avatar.h"
+#include "bodygraph.h"
+#include "bodypart.h"
+#include "cursesdef.h"
+#include "damage.h"
+#include "generic_factory.h"
+#include "input.h"
+#include "ui_manager.h"
+
+#define BPGRAPH_MAXROWS 20
+#define BPGRAPH_MAXCOLS 40
+
+static const bodygraph_id bodygraph_full_body( "full_body" );
+
+namespace
+{
+
+generic_factory<bodygraph> bodygraph_factory( "bodygraph" );
+
+} // namespace
+
+/** @relates string_id */
+template<>
+const bodygraph &string_id<bodygraph>::obj() const
+{
+    return bodygraph_factory.obj( *this );
+}
+
+/** @relates string_id */
+template<>
+bool string_id<bodygraph>::is_valid() const
+{
+    return bodygraph_factory.is_valid( *this );
+}
+
+void bodygraph::load_bodygraphs( const JsonObject &jo, const std::string &src )
+{
+    bodygraph_factory.load( jo, src );
+}
+
+void bodygraph::reset()
+{
+    bodygraph_factory.reset();
+}
+
+const std::vector<bodygraph> &bodygraph::get_all()
+{
+    return bodygraph_factory.get_all();
+}
+
+void bodygraph::finalize_all()
+{
+    bodygraph_factory.finalize();
+}
+
+void bodygraph::check_all()
+{
+    bodygraph_factory.check();
+}
+
+void bodygraph::load( const JsonObject &jo, const std::string & )
+{
+    optional( jo, was_loaded, "parent_bodypart", parent_bp );
+    optional( jo, was_loaded, "fill_sym", fill_sym );
+    if( jo.has_string( "fill_color" ) ) {
+        fill_color = color_from_string( jo.get_string( "fill_color" ) );
+    }
+
+    if( !was_loaded || jo.has_array( "rows" ) ) {
+        rows.clear();
+        for( const JsonValue jval : jo.get_array( "rows" ) ) {
+            if( !jval.test_string() ) {
+                jval.throw_error( "\"rows\" array must contain string values." );
+            } else {
+                std::wstring wrow = utf8_to_wstr( jval.get_string() );
+                std::vector<std::string> row;
+                for( const wchar_t wc : wrow ) {
+                    row.emplace_back( wstr_to_utf8( std::wstring( 1, wc ) ) );
+                }
+                rows.emplace_back( row );
```

Comment:
If you want to extract individual wide characters from the UTF-8 encoded string, there is a `utf8_display_split` function available.